src/libmeasurement_kit/ooni/telegram.cpp

```diff
@@ -0,0 +1,194 @@
+                }
+            }
+
+        };
+    };
+
+    for (auto ip_port : ip_ports) {
+        std::list<std::string> ip_port_l = split(ip_port, ":");
+        if (ip_port_l.size() != 2) {
+            logger->warn("Couldn't split ip_port: %s", ip_port.c_str());
+            // (*entry)["failure"] = ValueError().as_ooni_error();
+            // callback(entry);
+            return;
+        }
+        std::string ip = ip_port_l.front();
+        int port = std::stoi(ip_port_l.back());
+
+        Settings tcp_options;
+        tcp_options["host"] = ip;
+        tcp_options["port"] = port;
+        tcp_options["net/timeout"] = 10.0;
+        templates::tcp_connect(tcp_options, tcp_cb(ip, port), reactor, logger);
+    }
+
+}
+
+static void http_many(const std::vector<std::string> urls,
+                      Var<Entry> entry,
+                      Callback<Error> cb,
+                      Var<Reactor> reactor,
+                      Var<Logger> logger) {
+    int urls_count = urls.size();
+    Var<int> urls_tested(new int(0));
+    if (urls_count == *urls_tested) {
+        cb(NoError());
+        return;
+    }
+
+    auto http_cb = [=](std::string url) {
+        return [=](Error err, Var<http::Response> response) {
+             Entry result = {
+                 {"url", url},
+                 {"status",
+                     {{"success", nullptr},
+                     {"failure", nullptr}}},
+             };
+             if (!!err) {
+                 logger->info("telegram: failure HTTP connecting to %s",
+                     url.c_str());
+                 result["status"]["success"] = false;
+                 result["status"]["failure"] = err.as_ooni_error();
+             } else {
+                 logger->info("telegram: success HTTP connecting to %s",
+                     url.c_str());
+                 result["status"]["success"] = true;
+             }
+            (*entry)["http_connect"].push_back(result);
+             *urls_tested += 1;
+             if (urls_count == *urls_tested) {
+                 cb(NoError());
+             }
+         };
+    };
+
+    for (auto url : urls) {
+        Settings http_options;
+        http_options["http/url"] = url;
+        http::Headers headers = constants::COMMON_CLIENT_HEADERS;
+        std::string body;
+        templates::http_request(entry, http_options, headers, body,
+                                http_cb(url), reactor, logger);
+    }
+
+}
+
+void telegram(std::string input, Settings options,
+              Callback<Var<report::Entry>> callback,
+              Var<Reactor> reactor, Var<Logger> logger) {
+    std::vector<std::string>
+        TELEGRAM_WEB_URLS = { "http://web.telegram.org/",
+                              "https://web.telegram.org/" };
+    // should probably just make these std::pair<std::string,int>,
+    // but I'm not sure if this will be better later when I get
+    // rid of the duplication between this and the http ones
```

Comment:
You can probably write a function that maps the list of pairs to a list of URLs.

Reply:
Don't know what you mean?